src/main/flight/alt_hold.c

```diff
@@ -38,25 +38,20 @@ typedef struct {
     float ki;
     float kd;
     float kf;
-    float previousAltitude;
     float integral;
-} simplePid_t;
+} altHoldPid_t;
```

Comment:
Can we use the struct (and extend if needed) from position.c (altitudePidCoeffs_t) ?
Likewise with gps_rescue.

Reply:
You mean, use the values in the struct directly instead of writing them at some init time into the other file? I didn’t know if that was possible, but it would be great if it could be done that way.

Reply:
Need to use extern. @ledvinap

Reply:
Yes @ledvinap  that's a really good question - when to use extern, and when not, in a general sense.
Whether the considerations are different for values vs structs, and also whether they are different for for a 'static' struct like this pid one, where the values are set on initialisation only, vs a 'dynamic' struct such as the one holding altitude and altitudeDerivative.  I guess in these cases we would want a 'read-only' method so the client code can't mess with the incoming values.

Reply:
I'm afraid I don't fully understand the problem.
Location of struct declaration and definition is mostly convention. The idea is to keep some encapsulation and modularization of code.
Composition of structs is easy too (`struct  { altitudePidCoeffs_t coefs;  float integral; } altHoldPid_t'`), only with too complicated code, it may create problems with circular `#include`s.

Ideally, there would be pid.[ch] that encapsulates PID behavior. And all other code would use it.
But with all the heuristics and PID 'improvements', it becomes complicated. This started as simple, quite generic PID. But now there is setpoint feedforward, i-term 'relaxing', i-term limiting, etc. All these changes help in some situation and usually are not detrimental in others.